WIP: Initial support for IRC

postlund opened the pull request:
## Description:
This PR adds support for Internet Relay Chat (IRC) in Home Assistant. Short summary of what is supported:

* Multiple IRC networks
* Sensors displaying topic, last speaker, or user count in a channel
* Notify platform for sending messages to a channel or user
* Device tracker that can tell if a user is in a channel or not
* Compatible with some quirks in znc (especially log replay)

Most useful settings can be specified for each server, like nick, username, hostname, password, SSL, etc. I have mostly tested this using ZNC (attaching to my already running bouncer) but it works fine as stand-alone as well. So you can use Home Assistant as an IRC bot if you like.

I am currently quite satisfied with the feature set for an initial release, but there are more features that can be added in the future (raw message, nickserv, changing modes for users, actions like kick, etc). Error management is something I feel needs to be improved before merging as it's hard to know what went wrong if it doesn't work. Feel free to try it out, come with feedback and wish for features.

Marking this PR as WIP until I've worked out the quirks regarding error management. Maybe I find some other issues as well.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
Setting up a new connection looks like this:
```yaml
irc:
  - host: irc.quakenet.org
    port: 6667
    nick: admin
    ssl: true
    verify_ssl: false
    real_name: Spiderman
    username: batman
    password: mysupersecretandsecurepassword  # specify user:password if you use znc
    network: quakenet
    znc: true
    autojoin:
      - '#hass'
      - '#otherchannel'
    whitelist:
      - robin
      - alfred.*  # regexp allowed
    only_when_addressed: true
```
Only ``host``, ``nick`` and ``network`` are required (the other are optional or have default values) but I have included all settings for clarity. Try to be as restrictive as possible with the whitelist as all messages from the listed users will be sent as events on the event bus. So you might flood it with lots of events if your IRC network is busy. If ``only_when_addressed`` is set to ``true``, an event will only by fired if the bot is addressed, i.e. message line starts with "<nick>:". This can be quite useful when creating a bot.

**You must add "" or '' around a channel as # otherwise make the rest of the line into a comment!**

To add a new sensor, add this:
```yaml
sensor:
  - platform: irc
    network: quakenet
    channel: '#hass'
    attribute: last_speaker
```
This will create a sensor displaying the last message someone said in #hass. Note that the ``network`` (quakenet) must match ``network`` in the server description. The same principal is used for other platforms as well.

Supported values for ``attribute`` is: topic, last_speaker, users (you can probably figure out what they do).

To add a device tracker, simply add this:
```yaml
device_tracker:
  - platform: irc
    network: quakenet
    user: ballob
    channel: '#hass'
```
This will setup a device tracker that changes state to "online" or "offline" depending on if ballob is in #hass on quakenet or not. It works just like any other device tracker.

The notify platform is loaded automatically by the ``irc`` component, so you do not have to care about that. To send a message, just use ``notify.notify``:
```yaml
service: notify.notify
data:
  target: '#hass'
  message: Hello!
  data:
    network: quakenet
```
If ``network`` is not specified, the message will be broadcasted on all your networks. So it's safe to leave out if you only have one network, otherwise you should probably include it.

How to put this to good use? Lets create a very, very simple IRC bot. It will listen to what is said in #hass from users specified in the whitelist and if it matches an entity, it will display the state of that entity.

Now, lets define an automation. It will trigger every time a whitelisted users says something. Some simple jinja script checks if the specified entity exists and replies with a message containing the state (or "Unknown entity" if the entity does not exist):
```yaml
automation:
  - alias: Print entity states to IRC
    trigger:
      - platform: event
        event_type: irc_privmsg
        # event_data:
        #   channel: '#hass'
    action:
      service: notify.notify
      data_template:
        target: >
          {{ trigger.event.data['channel'] }}
        message: >
          {% set message = trigger.event.data['last_message'] %}
          {% set entity_id = message.split(' ')[1] %}
          {% set state = states(entity_id) %}

          {% if state != 'unknown' %}
            State: {{ state }}
          {% else %}
            Unknown entity: {{ entity_id }}
          {% endif %}
```
Just address the bot with some entity in the channel (e.g. ``mybot: sensor.outside``) and it will print its state. Neat! This is of course very simple, but it shows some of it's potential. Combine with ``python_script`` and it should be quite flexible.

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

postlund committed:
Initial support for IRC

pvizeli reviewed (changes requested):
The sensor `last_message` is not a real sensor. It have the same problem as smtp sensor. This data need to be a event inside event bus and not a sensor value. A sensor value can be a user count of the channel or other things, but a message need to fire a event and not be sensor state. While this can floot the event bus, you should change the logic from ignore user to whitelist users

postlund commented:
I see your point there @pvizeli and I tend to agree. As you say, it would flood the event bus if events are generated for each message but I would like to not limit the functionality to a specific set of users (although, it might be a good idea to add support for that). So, I propose another solution to this by removing ``last_message`` as a sensor and adding it as a state attribute when using ``last_speaker``. That way a sensor using ``last_speaker`` would trigger an update every time someone says something inside a channel and the message could be retrieved as an attribute. This would not flood the event bus and still allow for creating more generic IRC bots.

Would this be an acceptable solution?

pvizeli commented:
A state change (also only if atttribute have change) produce also a event on event bus. So you can fire a Event with include all this data inside Event.data - That allow you to make realy greate thins with template or scripts. The only way to stop flooting is with a withliste, that should be default on. So you can disabled it and know that you DB will growup a lot, but the default behave is with whilelist and user run not into this trouble per default.

postlund committed:
A few updatest to the IRC component

  * Sensor now only supports topic, last_speaker and users (number of
    users in channel) as attribute

  * Channel message are sent as an event on event bus (irc_privmsg)
    instead of being a sensor

  * Whitelist instead of blacklist for users (with regexp)

  * Notify excepts "network" as additional data

  * Additional hack to make irc3 dependency work correctly

postlund commented:
Thanks for the input @pvizeli! I agree with you and I have made the changes you requested and updated the description so that it matches. Also fixed a few other minor things, so hopefully it passes all the checks too.

postlund committed:
Ignore coverage for IRC

postlund committed:
Add flag to require bot to be addressed

If only_when_addressed is true, an irc_privmsg event will only be fired
if the bot is addressed with ("<nick>:" first at line) and user is on
the whitelist.

postlund commented:
I added an additional flag, ``only_when_addressed`` (default: true), that only fires an event if the bot is properly addressed by its nick. So a line must start with "nick_of_bot:", otherwise it's ignored. The whitelist is applied as before as well.

balloob commented on homeassistant/components/device_tracker/irc.py:
```diff
@@ -0,0 +1,65 @@
+
+            dev_id = slugify('{0} {1}'.format(user, channel))
+            status = 'online' if user in value else 'offline'
+            hass.async_add_job(
```
Is this in an async context? That means that whatever is calling the observer is async, but I did not see any annotations on the Observable class.

postlund replied:
Yes, ``attr_updated`` is called whenever some property change in the Observable class (effectively IRCCchannel ATM), which is basically done by the ``on_``-methods in HAPlugin. All of those methods are run in the event loop, so Observable is more or less a proxy-class here. We should consider all the ``on_``-methods as callback methods from async context, mainly due to how the irc3 library is designed. Is that clear enough? I feel that I'm not the best at explaining this...

balloob replied:
> Run in the event loop

That's all you needed to say to explain 👍

balloob commented on homeassistant/components/irc.py:
```diff
@@ -0,0 +1,303 @@
+    # Plugins must have an attribute which is not possible to add globally
+    # since Home Assistant installs dependencies on-the-fly. This adds it in
+    # runtime instead (sort of a hack).
+    global HAPlugin  # pylint: disable=invalid-name,global-variable-undefined
```
Why does this need to be global? It doesn't seemed to be used global.

Also, would it be possible to store it inside `hass.data`

postlund replied:
Yeah, this is a fun little hack 😄 IRC3 uses venusian (https://pypi.python.org/pypi/venusian) to implement plugin support. It's supposed to make life easier for developer by only requiring decorators for certain things. To create a new plugin, decorate the class with irc3.plugin and add the module path to "includes" in the irc config (see async_setup_irc_server method). When a connection is setup to a server, it will scan all modules specified in "includes" and look for classes with that decorator and instantiate them, i.e. adding a new plugin.

This "global" hack defines a new class at top level of the module after it has been loaded

balloob commented on homeassistant/components/sensor/irc.py:
```diff
@@ -0,0 +1,96 @@
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Handle when an entity is about to be added to Home Assistant."""
+        self._channel.observe(self)
```
Is `observe` async friendly? If so, should we prefix the name with `async_` ?

postlund replied:
``observer`` is only supposed to add an observer to an internal list of observers for later use, so it will never block or do any blocking I/O of any sort (definition is fixed). So I could stretch as far as @callback, but it seems weird to do that since it's not really a callback.

balloob commented:
So looks like this PR kinda fell off the radar, sorry about that. Left some comments.

postlund commented:
@balloob It's OK, I've had vacation and not worked on it for some time. I also wasn't expecting any review comments since I marked it with WIP, so it's all fine by me 😄

balloob commented on homeassistant/components/irc.py:
```diff
@@ -0,0 +1,303 @@
+                return
+
+            channel = self.get_channel(target)
+            self.hass.bus.fire(EVENT_PRIVMSG, {
```
If this is done inside the event loop, it should be async_fire.

balloob replied:
In fact, `bus.fire` should not even work when called from inside the event loop.

andrey-git commented:
Any updates on this PR?

postlund commented:
I'm working on some Apple TV stuff at the moment, so I currently don't put much time into this component. I will finish it when I get some time.

balloob commented:
@postlund going to close this, you can open a new PR when you're ready to finish it.
